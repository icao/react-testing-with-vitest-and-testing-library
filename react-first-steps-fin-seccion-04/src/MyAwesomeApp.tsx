import { CSSProperties } from "react";

const firstName = "Cesar";
const lastName = "Aparicio";

const favoriteGames = ["Elden Ring", "Smash", "Metal Gear"];
const isActive = false;

const address = {
  zipCode: "ABC-123",
  country: "CanadÃ¡",
};

const myStyles: CSSProperties = {
  backgroundColor: "#fafafa",
  borderRadius: 20,
  padding: 10,
  marginTop: 30,
  color: "#333",
  border: "1px solid #3c3c",
};

export const MyAwesomeApp = () => {
  return (
    <>
      <h1 data-testid="first-name-title"> {firstName} </h1>
      <h3> {lastName} </h3>
      <h3> {lastName} Aqui mas </h3>

      <p>{favoriteGames.join(", ")}</p>
      <p>{2 + 2}</p>

      <h1>{isActive ? "Activo" : "No activo"}</h1>

      <p style={myStyles}>{JSON.stringify(address)}</p>
    </>
  );
};

// export function MyAwesomeApp() {
//   return (
//     <>
//       <h1>Fernando</h1>
//       <h3>Herrera</h3>
//     </>
//   );
// }
