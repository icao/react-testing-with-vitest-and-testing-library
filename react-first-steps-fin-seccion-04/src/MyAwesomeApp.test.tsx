import { describe, expect, test } from "vitest";
import { render, screen } from "@testing-library/react";
import { MyAwesomeApp } from "./MyAwesomeApp";

describe("MyAwesomeApp", () => {
  test("Should render the first name and lastname", () => {
    // container solo sirve para mostrar el contenido de la pagina inicial. Pero si se ejecuta un evento que modifica el estado de la pagina, el contenido no se actualiza.
    const { container } = render(<MyAwesomeApp />);
    // console.log(screen.debug());
    const h1 = container.querySelector("h1");
    const h3 = container.querySelector("h3");

    expect(h1?.innerHTML).toContain("Cesar");
    expect(h3?.innerHTML).toContain("Aparicio");
  });

  test("Should render the first name and lastname - screen", () => {
    // screen.debug() sirve para mostrar el contenido de la pagina inicial y si se ejecuta un evento que modifica el estado de la pagina, el contenido se actualiza.
    render(<MyAwesomeApp />);
    // screen.debug();

    // const h1 = screen.getByRole('heading', { level: 1 }) // Encuentra todas las coincidencias, no es especifico

    const h1 = screen.getByTestId("first-name-title"); // Encuentra el elemento especifico pero se debe usar el atributo data, no es lo recomendable aunque es mas limpio.

    console.log("ðŸ‘‰", h1.innerHTML);

    expect(h1.innerHTML).toContain("Cesar");
  });

  test("should march Snapshot", () => {
    const { container } = render(<MyAwesomeApp />);
    // toMatchSnapshot() - crea una imagen del componente renderizado y la compara con una imagen guardada previamente. Si no existe la imagen guardada, la crea.
    // Si la imagen guardada y la nueva imagen son diferentes, el test falla.
    expect(container).toMatchSnapshot();

    // Si el codigo varia, el snapshot debe ser actualizdo si se quiere que el test pase.
  });

  test("should march Snapshot - screen", () => {
    render(<MyAwesomeApp />);

    expect(screen.getByTestId("my-awesome-app")).toMatchSnapshot();

    //!NOTA: No se recomienda usar el testid con. el snapshot porque este puede cambiar, tambien es muy volatil el testid, es por eso que no se suele usar.
    //! Tambien se debe tener cuidado con el snapshot si es que le codigo cambia.

  });
});

//! IMPORTANTE: Se recomienda usar:
//! Container: para testear sobre un contenido que no sera moridifado por un evento
//! Screen por si se quiere testear sobre un contenido que sera moridifado por un evento
//! Testid: solo en casos muy especificos, no es lo recomendable