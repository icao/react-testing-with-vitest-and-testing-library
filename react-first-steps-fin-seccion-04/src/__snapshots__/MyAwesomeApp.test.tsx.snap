// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyAwesomeApp > should march Snapshot - screen 1`] = `
<section
  data-testid="my-awesome-app"
>
  <h1
    data-testid="first-name-title"
  >
     
    Cesar
     
  </h1>
  <h3>
     
    Aparicio
     
  </h3>
  <h3>
     
    Aparicio
     Aqui mas 
  </h3>
  <p
    class="mi-clase-favorita"
  >
    Elden Ring, Smash, Metal Gear
  </p>
  <p>
    4
  </p>
  <h1>
    No activo
  </h1>
  <p
    style="background-color: rgb(250, 250, 250); border-radius: 20px; padding: 10px; margin-top: 30px; color: rgb(51, 51, 51);"
  >
    {"zipCode":"ABC-123","country":"Canadá"}
  </p>
</section>
`;

exports[`MyAwesomeApp > should march Snapshot 1`] = `
<div>
  <section
    data-testid="my-awesome-app"
  >
    <h1
      data-testid="first-name-title"
    >
       
      Cesar
       
    </h1>
    <h3>
       
      Aparicio
       
    </h3>
    <h3>
       
      Aparicio
       Aqui mas 
    </h3>
    <p
      class="mi-clase-favorita"
    >
      Elden Ring, Smash, Metal Gear
    </p>
    <p>
      4
    </p>
    <h1>
      No activo
    </h1>
    <p
      style="background-color: rgb(250, 250, 250); border-radius: 20px; padding: 10px; margin-top: 30px; color: rgb(51, 51, 51);"
    >
      {"zipCode":"ABC-123","country":"Canadá"}
    </p>
  </section>
</div>
`;
